docs/source/blogs/tech_blog/blog12_Inference_Time_Compute_Implementation_in_TensorRT-LLM.md

```diff
@@ -0,0 +1,425 @@
+
+            # For the probe task, append the suffix to force a chain-of-thought leading to an answer.
+            yield [proposer_task, probe_task]
+
+            # Retrieve the output from the probe task.
+            probe_text = probe_task.output_str
+
+            # Extract the potential answer from the probe response.
+            answer = self.obtain_answer(probe_text)
+            probe_answers.append(answer)
+            probe_responses.append(probe_text)
+
+            if self.should_early_stop(probe_answers, probe_responses):
+                tasks[0].result = probe_task.result
+                # If the current prompt indicates the chain-of-thought phase has ended, use one type of suffix.
+                if "</think>" in current_prompt:
+                    tasks[0].output_str = (current_prompt + self.answer_suffix +
+                                        probe_answers[-1] + "}\n\\]")
+                    return
+                else:
+                    # Otherwise, use the suffix with marker to transition clearly.
+                    tasks[0].output_str = (current_prompt +
+                                        self.answer_suffix_with_marker +
+                                        probe_answers[-1] + "}\n\\]")
+                    return
+
+            # If not confident, do another round of generation
+            # Append the newly generated text from the proposer to the current prompt for the next iteration.
+            current_prompt += proposer_task.output_str
+
+        # If the maximum token limit is reached without satisfying the certainty condition,
+        # output the accumulated prompt as the final output.
+        tasks[0].result = proposer_task.result
+        tasks[0].output_str = current_prompt
+        return
+```
+The `probe_task` can utilize prefix kvcache reuse to enhance inference performance. TensorRT LLM enables the kvcache of an in-progress request to be reused by other requests, so `probe_task` can `proposer_task`'s kvcache even though the `proposer_task` is in a continuous running state.
+
+Now we have implemented a `Controller` for Dynasor-CoT. Here is an example of how to use it:
+```python
+dynasor_generation_controller = DynasorGenerationController(
+    # Parameters for DynasorGenerationController
+    )
+
+llm = ScaffoldingLlm(
+    prototype_controller=dynasor_generation_controller, 
+    # other parameters for ScaffoldingLLM
+    )
+results = llm.generate(prompts)
+```
+
+### Implement Dynasor-CoT based Majority Voting in Scaffolding
+Scaffolding is designed to be modular and reusable. We can assemble methods just like LEGO building blocks. For instance, to implement Dynasor-CoT-based Majority Voting, we can simply stack our `DynasorGenerationController` with a `MajorityVoteController`.
+
+Once a controller for majority voting is built, no further implementation is needed. We can directly stack the two controllers as shown below.
```

Comment:
suggestion: Once a controller for majority voting is built, no additional implementation is required.